packages/theme-base/sd.config.js

```diff
@@ -72,3 +77,26 @@ function CommonJSNestedFormatter({ dictionary, options, file }) {
     fileHeader({ file }) + `module.exports = ${JSON.stringify(theme, null, 2)};`
   );
 }
+
+function ThemeTypeFormatter({ dictionary, file }) {
+  const { fileHeader } = formatHelpers;
+  return (
+    fileHeader({ file }) +
+    `export interface AmplifyTheme {${ThemeToTypes(dictionary.tokens, 1)}\n}`
+  );
+}
+
+function ThemeToTypes(obj, depth = 0) {
+  let output = ``;
+  for (const key in obj) {
+    const value = obj[key];
+    const indent = [...Array(depth).keys()].map(() => `  `).join('');
+    output += `\n${indent}${key}?: `;
```

Comment:
My understanding was that the entirety of the theme object was required to exist for styles to computed correctly.

If the `<AmplifyProvider theme={...}>` value is a subset, we could remove the optional `?:` from the generation and instead try `Partial<AmplifyTheme>` within `<AmplifyProvider>`:

> https://www.typescriptlang.org/docs/handbook/utility-types.html#partialtype

(If my understanding is wrong an it's true that every property of a theme, even outside of the provider, is optional, then ignore me!)

Reply:
This might be something worth discussing more broadly. I guess this is more of a `ThemeOverrides` type that a user could supply than an `AmplifyTheme` type. Based on @zchenwei's theme provider PR, a user could supply a theme as simple as:

```javascript
{
  colors: {
    font: {
      primary: "blue"
    }
  }
}
```

And it would just override that one design token in the theme. Although at the provider level would we want to enforce it is given a valid Theme that has at least all of the design tokens in our default theme defined? 

TIL about `Partial` in typescript, that looks like exactly what we want. Does it apply recursively for a deeply nested object?

Reply:
As far as I know, it is a shallow `Partial` but making a recursive one should be simple. I've have do that earlier actually, but I remove it since not sure if we should strictly type the object as `Theme`